### PR TITLE
lifted decision list data structures

### PR DESCRIPTION
**Pull request by tomsilver**
starting to work towards reimplementing PG3 in this code base.

### CODE REVIEW

File: src/structs.py [hunk 1]
```diff
@@ -1333,6 +1333,144 @@ class PathToStateResponse(Response):
...
+        assert isinstance(other, _GroundLDLRule)
+        return str(self) > str(other)
+
+
+@dataclass(frozen=True)
+class LiftedDecisionList:
+    """A goal-conditioned policy from abstract states to ground NSRTs
+    implemented with a lifted decision list.
+
+    Given an abstract state and goal, the rules are grounded in order. The
```
Comment (ronuchit) [inline src/structs.py:1463]: get rid of this para and the next? it's redundant with `utils.query_ldl` and it's also confusing bc this isn't a function

**Review by ronuchit — APPROVED** [review]

File: src/utils.py [hunk 2]
```diff
@@ -1709,6 +1710,15 @@ def all_possible_ground_atoms(state: State,
     return sorted(ground_atoms)
 
 
+def all_ground_ldl_rules(
+        rule: LDLRule,
+        objects: Collection[Object]) -> Iterator[_GroundLDLRule]:
+    """Get all possible groundings of the given NSRT with the given objects."""
```
Comment (ronuchit) [inline src/utils.py:1716]: update this comment ("NSRT")

File: src/utils.py [hunk 3]
```diff
@@ -2439,3 +2449,21 @@ def nostdout() -> Generator[None, None, None]:
     sys.stdout = _DummyFile()
     yield
     sys.stdout = save_stdout
+
+
+def query_ldl(ldl: LiftedDecisionList, atoms: Set[GroundAtom],
+              goal: Set[GroundAtom]) -> Optional[_GroundNSRT]:
```
Comment (ronuchit) [inline src/utils.py:2455]: why do we return the ground NSRT instead of the ground rule?

Comment (tomsilver) [reply src/utils.py:2455]: I'm thinking about this like querying a (partial) policy in the abstract space. also, I don't foresee the rule being useful except to get the ground NSRT from it

**Pull request merged**